ramips: convert legacy SoCs wireless calibration EEPROM to NVMEM format

DragonBluep opened the pull request:
Notice:
1. All NIC models are judged based on the commit message or the makefile default packages.
2. All Ralink RTxxxx SoCs use the same rt2800 wireless driver, the EEPROM size is the same 0x200.
3. MT7620 build-in WiFi is driven by rt2800 so the EEPROM size is the same 0x200.
4. MT7628 build-in WiFi is driven by kmod-mt7603 so the EEPROM size is the same 0x400.
5. Only three devices use MT7613 PCIe NIC, see commit message.
6. The remaining MT76 PCIe NIC are all MT7610 or MT7612, EEPROM size is 0x200.
7. MT76 devices use property `mediatek,mtd-eeprom` and Ralink devices use `ralink,mtd-eeprom`.

|Package Name|NIC Model|EEPROM size||
|-|-|-|-|
|kmod-mt76x0e|MT7610|0x200|same as MT7612|
|kmod-mt76x2|MT7602/MT7612|0x200||
|kmod-mt7663-firmware-ap|MT7613|0x4da8||

Ansuel commented:
this was done manually or with an automated script?

Ansuel commented:
@DragonBluep I notice you are putting lots of stuff here... I would split this in more pr
- simple pr for device that are missing a package (and eventually if needed the missing node)
- pr making the conversion from mediatek special binding to nvmem
- pr migrating and introducing the new patch for any other device

Is it problematic for you? Splitting this would make this easier to review and would also make part of this pr to be merged faster.

DragonBluep committed:
ramips: convert rt2x00 EEPROM to NVMEM format

This patch converts legacy Ralink SoCs and MT7620 WiFi calibration
data to NVMEM format. The EEPROM size is 0x200.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: convert MT7628 EEPROM to NVMEM format

This patch converts MT7628 WiFi calibration data to NVMEM format. The
EEPROM size is 0x400.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: convert mt76 PCIe NIC EEPROM to NVMEM format for legacy SoCs

This patch converts MT761{0,2,3} PCIe WiFi calibration data to NVMEM
format for legacy Ralink SoCs (MT7620 and Mt7628). The EEPROM size of
the MT7610 and MT7612 is 0x200. there are only three devices uses
MT7613 NIC, ASUS RT-AC1200 V2, COMFAST CF-WR758AC V2 and Keenetic
KN-1613. The EEPROM size of them is 0x4da8.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
> this was done manually or with an automated script?

I manually edited these things. Because different NICs have different EEPROM sizes.

> @DragonBluep I notice you are putting lots of stuff here... I would split this in more pr
> 
> * simple pr for device that are missing a package (and eventually if needed the missing node)
> * pr making the conversion from mediatek special binding to nvmem
> * pr migrating and introducing the new patch for any other device
> 
> Is it problematic for you? Splitting this would make this easier to review and would also make part of this pr to be merged faster.

I split them into two PRs.

This PR converts the EEPROM of the legacy SoCs into NVMEM. It is still big but should be easily reviewed. We only need to focus on the three MT7613 devices because all other devices have the same EEPROM size `0x200`. They can be safely merged.

Another PR focuses on converting some MT7621 devices. The reviewer needs to refer to the device package on the MT7621 makefile or the wiki.

All other small fixes will be sent separately to the mailing list.

DragonBluep committed:
ramips: fix DTS EEPROM property for some MT7628 devices

The MT7628 integrated wireless is driven by `mt76`, so the right
EEPROM property name is `mediatek,mtd-eeprom`.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: convert the remaining mtd-mac-address to NVMEM format

`mtd-mac-address` has been abandoned. Therefore, convert them to
NVMEM format. This patch also removes some useless mtd-mac-address
properties.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: add missing mt76 packages for Telco Electronics X1

Telco Electronics X1 has MT7603E and MT7612E PCIe NICs. They are
driven by kmod-mt7603 and kmod-mt76x2.

Ref: 73e0f52b6e4e ("ramips: add support for Telco Electronics X1")
Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: fix mtd partition node names for Phicomm PSG1208

The mtd partition node name should be "partition@${offset}".
However, the offsets of the PSG1208 don't match the partition
'reg' properties. This patch correct the wrong offsets.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

musashino205 commented:
```nvmem-layout``` may be necessary for newer binding.

DragonBluep commented:
> `nvmem-layout` may be necessary for newer binding.

`nvmem-layout` seems to only be used to provide basic MAC address support. I'm not sure if we really need it. On the other hand, introducing it into this PR may not be a good idea, as it will make review more difficult.

Ansuel commented on package/kernel/mac80211/patches/rt2x00/605-wifi-rt2x00-support-loading-eeprom-from-NVMEM-ce.patch:
```diff
@@ -0,0 +1,87 @@
++
++	rt2x00dev->eeprom = (__le16 *)data;
++
++	if (of_find_property(np, "ralink,eeprom-swap", NULL))
++		for (i = 0; i < len/sizeof(u16); i++)
++			rt2x00dev->eeprom[i] = swab16(rt2x00dev->eeprom[i]);
```
Is this actually used? Can't find it used in the current conversion.

Ansuel replied:
Notice this same thing is for the mtd... Do we use the mtd properties? 

In any case I would generalize the code and unify the properties... (for the conversion a simple grep and sed should do the trick)

DragonBluep replied:
> Is this actually used? Can't find it used in the current conversion.

Yes, some lantiq devices need it.
https://git.openwrt.org/?p=openwrt%2Fopenwrt.git&a=search&h=HEAD&st=grep&s=ralink%2Cmtd-eeprom-swap%3B

BTW, I think we can change the mtd eeprom patch property name to the same one `ralink,eeprom-swap`.
https://github.com/openwrt/openwrt/blob/b35756446391d1b21970a0e05f8f6f5e069f73cf/package/kernel/mac80211/patches/rt2x00/604-rt2x00-load-eeprom-on-SoC-from-a-mtd-device-defines-.patch#L85

Ansuel commented:
@DragonBluep I'm checking the mtd eeprom code and did it work? I'm seeing various problems in the implementation... Example ee->size is never set in the mtd function resulting in the function ALWAYS rejecting mtd eeprom??? Am I missing something?

I have some comments for your patch but honestly i'm more keen on reorganizing the eeprom code.

Ansuel commented:
Also why the eeprom code was never proposed upstream?

Ansuel commented on target/linux/ramips/dts/rt5350_olimex_rt5350f-olinuxino.dtsi:
```diff
@@ -70,7 +81,9 @@
-	ralink,mtd-eeprom = <&factory 0x0>;
+	nvmem-cells = <&eeprom_factory_0>;
+	nvmem-cell-names = "eeprom";
+
```
Remove extra space.

Ansuel commented on target/linux/ramips/dts/rt5350_poray_m4.dtsi:
```diff
@@ -92,16 +103,8 @@
-	macaddr_factory_4: macaddr@4 {
-		reg = <0x4 0x6>;
-	};
+	ralink,led-polarity = <1>;
```
Move this up.

Ansuel commented on target/linux/ramips/dts/rt5350_poray_x5.dts:
```diff
@@ -132,16 +143,8 @@
-	macaddr_factory_4: macaddr@4 {
-		reg = <0x4 0x6>;
-	};
+	ralink,led-polarity = <1>;
```
Move this up.

Ansuel commented on target/linux/ramips/dts/rt5350_poray_x8.dts:
```diff
@@ -93,16 +104,8 @@
-	macaddr_factory_4: macaddr@4 {
-		reg = <0x4 0x6>;
-	};
+	ralink,led-polarity = <1>;
```
Move this up.

Ansuel commented on target/linux/ramips/dts/rt5350_tenda_3g150b.dts:
```diff
@@ -110,16 +121,7 @@
 
 &wmac {
-	ralink,mtd-eeprom = <&factory 0x0>;
+	nvmem-cells = <&eeprom_factory_0>;
+	nvmem-cell-names = "eeprom";
```
This down.

Ansuel commented on target/linux/ramips/dts/rt5350_zyxel_keenetic-4g-b.dts:
```diff
@@ -121,16 +132,8 @@
-	macaddr_factory_28: macaddr@28 {
-		reg = <0x28 0x6>;
-	};
+	ralink,led-polarity = <1>;
```
Move this up.

Ansuel reviewed (changes requested):
![image](https://github.com/openwrt/openwrt/assets/20289090/1ea550e1-2ff1-4dc4-b7d4-8856cc3e78e2)

Hi, another session of paper,please. But this time this was exhausting AHAHAH

Anyway, lots of nitpick but I found some error to fix and some additional fix to do.

Please don't rebase (ignore the merge conflict) and push this only with the requested change so I can compare them.

We can care later on fixing the conflict but please don't rebase on top of main so I can track changes!
(and also ignore the comments about the nvmem patch, lets first address the DT stuff)

DragonBluep committed:
ramips: remove useless status property from wmac dts node

On the ramips target, all 'wmac' nodes in SoC dtsi are enabled by
default except mt7628. There is no need to mark them as 'okay'
again. So these useless properties can be removed.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: fix frequency limit property for some Edimax devices

These three devices uses MT7612E PCIe NIC and supported by the
'mt76' driver. So the right frequency limit property should be
`ieee80211-freq-limit`.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep committed:
ramips: fix factory partition size for SNR-CPE-W4N

The starting address of 'factory' partition is 0x40000, and the
starting address of the next partition is 0x50000. It's obvious
that the correct size for the 'factory' is 0x10000, just like
other MT7620 devices.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented on target/linux/ramips/dts/mt7620a_buffalo_whr-300hp2.dts:
```diff
@@ -146,18 +157,10 @@
-	ralink,mtd-eeprom = <&factory 0x0>;
+	nvmem-cells = <&eeprom_factory_0>;
+	nvmem-cell-names = "eeprom";
+
```
Now all similar nodes with both 'pinctrl' and 'nvmem' properties get the same format, 'pinctrl' first and then 'nvmem'.

DragonBluep commented on target/linux/ramips/dts/mt7620a_edimax_br-6478ac-v2.dts:
```diff
@@ -193,8 +209,9 @@
 	wifi@0,0 {
 		reg = <0x0000 0 0 0 0>;
-		mediatek,mtd-eeprom = <&factory 0x8000>;
-		mediatek,2ghz = <0>;
```
Done in https://github.com/openwrt/openwrt/pull/13584/commits/5984965f0f59112e67c5bf0bfd30250a5312660e.

DragonBluep commented on target/linux/ramips/dts/mt7620a_edimax_ew-747x.dtsi:
```diff
@@ -174,9 +189,8 @@
-	ralink,mtd-eeprom = <&factory 0x0>;
-	nvmem-cells = <&macaddr_factory_4>;
-	nvmem-cell-names = "mac-address";
+	nvmem-cells = <&eeprom_factory_0>;
```
I added it back.

DragonBluep commented on target/linux/ramips/dts/mt7620a_iodata_wn-ac733gr3.dts:
```diff
@@ -105,15 +105,33 @@
+				compatible = "nvmem-cells";
 				label = "factory";
-				reg = <0x40000 0x8000>;
+				reg = <0x40000 0x10000>;
```
copy&paste issue, fixed. Thanks.

DragonBluep commented on target/linux/ramips/dts/mt7620a_phicomm_psg1208.dts:
```diff
@@ -67,19 +67,34 @@
 			};
 
-			partition@20000 {
+			partition@30000 {
```
I’ve sent some small patches to the mailing list some days ago. Anyway, I pulled them back to this PR. https://github.com/openwrt/openwrt/pull/13584/commits/65ef9ec27a166310eb433416837e90060a2c5637

>All other small fixes will be sent separately to the mailing list:
>https://patchwork.ozlabs.org/project/openwrt/list/?submitter=86477

DragonBluep commented on target/linux/ramips/dts/mt7620n_snr_cpe-w4n-mt.dts:
```diff
@@ -91,9 +91,16 @@
+				compatible = "nvmem-cells";
 				label = "Factory";
-				reg = <0x40000 0x100000>;
+				reg = <0x40000 0x10000>;
```
Fixed in a separate commit https://github.com/openwrt/openwrt/pull/13584/commits/f97e8723c387f9b1f5cf3ed84301a7dab6b983b5.

DragonBluep commented on target/linux/ramips/dts/mt7628an_asus_rt-ac1200.dtsi:
```diff
@@ -128,7 +132,8 @@
 		compatible = "mediatek,mt76";
 		reg = <0x0000 0 0 0 0>;
-		mediatek,mtd-eeprom = <&factory 0x8000>;
+		nvmem-cells = <&eeprom_factory_8000>;
```
I changed them in the new patch version, should be fine now but still need review.

DragonBluep commented on target/linux/ramips/dts/rt3662_dlink_dir-645.dts:
```diff
@@ -136,8 +147,10 @@
+	nvmem-cells = <&eeprom_factory_0>;
+	nvmem-cell-names = "eeprom";
+
 	ralink,5ghz = <0>;
```
I'm not sure if frequency limit is working on rt2x00 driver. They are supported by a local patch.
https://github.com/openwrt/openwrt/blob/main/package/kernel/mac80211/patches/rt2x00/608-rt2x00-allow_disabling_bands_through_dts.patch

DragonBluep commented on target/linux/ramips/dts/mt7620a_edimax_br-6208ac-v2.dts:
```diff
@@ -213,7 +217,8 @@
-		ieee80211-freq-limit = <5000000 6000000>;
+		nvmem-cells = <&eeprom_factory_8000>;
+		nvmem-cell-names = "eeprom";
+		mediatek,2ghz = <0>;
```
Wrong here, will be fixed later.

DragonBluep commented on target/linux/ramips/dts/mt7620a_cameo_810.dtsi:
```diff
@@ -163,10 +167,9 @@
+		nvmem-cells = <&eeprom_factory_8000>, <&macaddr_factory_28>;
+		nvmem-cell-names = "eeprom", "mac-address";
 		mac-address-increment = <2>;
+		ieee80211-freq-limit = <5000000 6000000>;
```
If it can be accepted, I would like to move all `ieee80211-freq-limit` to the top of the nvmem stuff. It looks a bit ugly.

Ansuel replied:
It's ok but please make this change in a different commit to reduce patch delta. (before the conversion)

DragonBluep commented:
> @DragonBluep I'm checking the mtd eeprom code and did it work? I'm seeing various problems in the implementation... Example ee->size is never set in the mtd function resulting in the function ALWAYS rejecting mtd eeprom??? Am I missing something?

I've tested it on MT7620 for several weeks, yes it works. The eeprom_size is initialized by the driver.
https://elixir.bootlin.com/linux/v6.5.6/source/drivers/net/wireless/ralink/rt2x00/rt2800soc.c#L225

> 
> I have some comments for your patch but honestly i'm more keen on reorganizing the eeprom code.
> 
> (Also still analyzing but to me it seems we are leaking the allocated data, also for your code... but to fix this I need to reorg even more thing...) (In short I can be very wrong but currently for anything that is not passed as a file in userspace, we are leaking data, doesn't grow but still never freed)

`rt2x00dev->eeprom` seems just an uninitialized pointer so we shouldn't free the heap returned by `nvmem_cell_read()` in this function. It should be freed when the radio is done.

Anyway, this part also looks wired for me, I just copied these code from another mtd eeprom patch. Yes, you are right, we need a close look.

DragonBluep commented:
free eeprom memory:
https://elixir.bootlin.com/linux/v6.5.6/source/drivers/net/wireless/ralink/rt2x00/rt2x00pci.c#L30
https://elixir.bootlin.com/linux/v6.5.6/source/drivers/net/wireless/ralink/rt2x00/rt2x00soc.c#L28
https://elixir.bootlin.com/linux/v6.5.6/source/drivers/net/wireless/ralink/rt2x00/rt2x00usb.c#L766

will be called when removing device:
https://elixir.bootlin.com/linux/v6.5.6/source/drivers/net/wireless/ralink/rt2x00/rt2x00soc.c#L120

It seems that we don't need to worry about memory leak. Maybe we should allocate memory for `rt2x00dev->eeprom` and then copy nvmem cell data to it.

Ansuel commented:
@DragonBluep as I tought... we are leaking memory... If you want I can take care of updating the mtd patch and then you can followup following my changes? 

btw i was talking to the eeprom_file that require data and size to be set to be correctly freed but from what I can see size value is not set...

Ansuel commented on target/linux/ramips/dts/mt7620a_netcore_nw5212.dts:
```diff
@@ -115,8 +126,10 @@
 &ethernet {
-	mtd-mac-address = <&factory 0x28>;
 	mediatek,portmap = "llllw";
+
```
Extra space? (don't hate me)

DragonBluep replied:
I'm happy to see these suggestions. Thank you!

Ansuel commented on target/linux/ramips/dts/mt7628an_asus_rt-ac1200.dtsi:
```diff
@@ -73,13 +73,22 @@
+					reg = <0x0 0x400>;
+				};
+
+				/* v2 eeprom size is 0x4da8, 'reg'
+				   will be overridden in v2 dts. */
```
Think it's better to add this in the v2 dts instead of here in the dtsi. Also for multiline comments use 

```
/*
 *
 */
```

Ansuel commented on target/linux/ramips/dts/mt7628an_comfast_cf-wr758ac.dtsi:
```diff
@@ -84,9 +85,26 @@
+					reg = <0x0 0x400>;
+				};
+
+				/* v2 eeprom size is 0x4da8, 'reg'
+				   will be overridden in v2 dts. */
```
Same comment here.

DragonBluep commented:
> @DragonBluep as I tought... we are leaking memory... If you want I can take care of updating the mtd patch and then you can followup following my changes?

Thanks, please fix that patch and maybe backport it to the stable branch. 

> 
> btw i was talking to the eeprom_file that require data and size to be set to be correctly freed but from what I can see size value is not set...

I saw it now. I'll add the file size to the patch.
https://github.com/openwrt/openwrt/blob/b35756446391d1b21970a0e05f8f6f5e069f73cf/package/kernel/mac80211/patches/rt2x00/602-rt2x00-introduce-rt2x00eeprom.patch#L258-L260

Ansuel commented:
> I saw it now. I'll add the file size to the patch.

nha please wait for the changed patch I will propose.

DragonBluep commented:
> > I saw it now. I'll add the file size to the patch.
> 
> nha please wait for the changed patch I will propose.

Okay. I can help to test it. I'll force push the PR tomorrow or later.

Ansuel commented:
@DragonBluep can you rebase this? Also think also here we need to revert some commit?

DragonBluep commented:
> @DragonBluep can you rebase this? Also think also here we need to revert some commit?

I pushed it twice. First push apply the suggested changes and the second one rebase it to the main branch to solve some conflicts on MT7621 dts files.

All devices in this PR are WiFi4/5 routers. None of them use MT7915 NIC so we don't need to revert anything.
